ops/main.py

```diff
@@ -208,27 +251,27 @@ def main(charm_class):
     try:
         charm = charm_class(framework, None)
 
-        if not has_dispatch:
+        if not juju_event.is_dispatch:
             # When a charm is force-upgraded and a unit is in an error state Juju
             # does not run upgrade-charm and instead runs the failed hook followed
             # by config-changed. Given the nature of force-upgrading the hook setup
             # code is not triggered on config-changed.
             #
             # 'start' event is included as Juju does not fire the install event for
             # K8s charms (see LP: #1854635).
-            if (juju_event_name in ('install', 'start', 'upgrade_charm')
-                    or juju_event_name.endswith('_storage_attached')):
+            if (juju_event.name in ('install', 'start', 'upgrade_charm')
+                    or juju_event.name.endswith('_storage_attached')):
```

Comment:
Does it make sense to pull this into the same class?

Reply:
Yes, to me it does. Some more of the helpers could also get moved in. In another PR perhaps.